### DISCUSSION TITLE
Trying to create a scatter plot of the lap number a driver pits on versus qualifying position

### DISCUSSION THREAD

**Discussion by laurence9899** [theOehrly/Fast-F1#452] [Q&A] [answered]
Hi,

I'm trying to create a scatter plot of a driver the lap number a driver pits versus their qualifying position. Here's the code I've so far. I know it's definitely wrong but I've really been trying to do this with ages and haven't make much progress. Any help is greatly appreciated. I have tried so many variations the code below may be completely wrong and I apologise of that.
Thank you

#Set up necessary imports
import fastf1
import fastf1.plotting as ff1p
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from fastf1.ergast import Ergast

#Set up cache for race data from fastf1
fastf1.Cache.enable_cache('')
ergast = Ergast()
year=2021
race=1
    
race = fastf1.get_session(year, race , 'R')
quali=fastf1.get_session(year,race,'Q')
race.load()
quali.load()    
quali_pos=quali.quali.results
print(quali_pos)
pitstops=ergast.get_pit_stops(year,race)
fig, ax = plt.subplots(figsize=(8, 8))

sns.scatterplot(data=quali_pos,
            x="LapNumber",
            y="quali_pos",
            ax=ax,,
            s=80,
            linewidth=0,
            legend='auto')

**Comment by Casper-Guo** [comment, marked as answer]
I try to work with Pandas over Ergast where possible. Here is my implementation:

```
import numpy as np
import pandas as pd
import fastf1 as f
import fastf1.plotting as p
import matplotlib.pyplot as plt
import seaborn as sns

race = f.get_session(2021, 1, 'R')
quali = f.get_session(2021, 1, 'Q')
race.load()
quali.load()

p.setup_mpl()

race_laps = race.laps
in_laps = race_laps[~race_laps["PitInTime"].isna()]

# needed because you are working with old races
# p.driver_color won't be satisfactory
# many different implementations possible
def get_driver_color(driver:str) -> str:
    if driver in p.DRIVER_TRANSLATE:
        return p.driver_color(driver)
    else:
        return "#FFFFFF"

fig, ax = plt.subplots()
num_laps = race_laps["LapNumber"].max()

# assumption: quali.results is ordered correctly
for (pos, driver) in enumerate(quali.results["Abbreviation"]):
    driver_pits = in_laps[in_laps['Driver']==driver]["LapNumber"].unique()
    sns.scatterplot(x=driver_pits,
                    y=np.full(shape=(len(driver_pits)), fill_value=pos+1),
                    c=get_driver_color(driver),
                    label=driver)
    plt.annotate(text=driver, xy=(num_laps, pos+1))

plt.xlim((1, num_laps))
plt.yticks([1, 5, 10, 15, 20])
ax.get_legend().remove()
ax.invert_yaxis()
plt.ylabel("Qualifying Position")
plt.xlabel("Lap Number")
plt.show()
```

Not the most pretty but it shows the basic idea.

![image](https://github.com/theOehrly/Fast-F1/assets/89810860/e0242e4e-3ebc-4beb-8cd8-1930091d90ea)

**Reply by laurence9899** [reply]
Thank you very much, yes I found Pandas much easier to use so far.
